src/controller.rs

```diff
@@ -0,0 +1,257 @@
+    ///
+    /// Only objects that have an `OwnerReference` for our main resource type will trigger
+    /// a reconciliation.
+    pub fn owns<Child: Clone + Meta + DeserializeOwned + Send + 'static>(
+        mut self,
+        api: Api<Child>,
+        lp: ListParams,
+    ) -> Self {
+        self.kube_controller = self.kube_controller.owns(api, lp);
+        self
+    }
+
+    /// Call this method once your Controller object is fully configured.
+    /// It'll start talking to Kubernetes and will call the `Strategy` implementation.
+    pub async fn run<S>(self, client: Client, strategy: S)
+    where
+        S: ControllerStrategy<Item = T> + 'static,
+    {
+        let context = Context::new(ControllerContext { client, strategy });
+
+        self.kube_controller
+            .run(reconcile, error_policy, context)
+            .for_each(|res| async move {
+                match res {
+                    Ok(o) => info!("Reconciled {:?}", o),
+                    Err(ref e) => error!("Reconcile failed: {:?}", e),
+                };
+            })
+            .await
+    }
+}
+
+/// The context used internally in the Controller which is passed on to the `kube_runtime::Controller`.
+/// Note that we can only get immutable references to this object so should we need mutability we need to model it as interior mutability (e.g. with a Mutex)
+struct ControllerContext<S>
+where
+    S: ControllerStrategy,
+{
+    client: Client,
+    strategy: S,
+}
+
+/// This method contains the logic of reconciling an object (the desired state) we received with the actual state.
+async fn reconcile<S, T>(
+    resource: T,
+    context: Context<ControllerContext<S>>,
+) -> Result<ReconcilerAction, crate::error::Error>
+where
+    T: Clone + Debug + DeserializeOwned + Meta + Send + Sync + 'static,
+    S: ControllerStrategy<Item = T> + 'static,
+{
+    let context = context.get_ref();
+
+    let client = &context.client;
+    let strategy = &context.strategy;
+
+    if handle_deletion(&resource, client.clone(), &strategy.finalizer_name()).await?
+        == ReconcileFunctionAction::Done
+    {
+        return Ok(create_non_requeuing_reconciler_action());
+    }
+
+    add_finalizer(&resource, client.clone(), &strategy.finalizer_name()).await?;
+
+    let rc = ReconciliationContext::new(context.client.clone(), resource.clone());
+
+    let state = strategy.init_reconcile_state(rc);
+    let futures = state.reconcile_operations();
+
+    for future in futures {
+        let result = future.await;
+
+        match result {
+            Ok(ReconcileFunctionAction::Continue) => {
+                trace!("Reconciler loop: Continue");
+            }
+            Ok(ReconcileFunctionAction::Done) => {
+                trace!("Reconciler loop: Done");
+                break;
+            }
+            Ok(ReconcileFunctionAction::Requeue(duration)) => {
+                trace!(?duration, "Reconciler loop: Requeue");
+                return Ok(ReconcilerAction {
```

Comment:
`create_requeuing_reconciler_action` takes a integer and not a duration.
I created the method because I didn't always want to create a `Duration` but that probably makes it much clearer so I guess I'll change it.